feat: Add cosmovisor --help

dwedul-figure opened the pull request:
## Description

Closes: #10126 

Adds `help` to cosmovisor.

Help will be printed if:
* Any argument is `help`, `--help`, or `-h`.
* Either of the `DAEMON_NAME` or `DAEMON_HOME` environment variables are not set.

The help text includes a description, info about the environment variables, and a link for more info.
It then checks the environment variables. If there are any errors with them, they are printed and execution is done; otherwise the configured values are printed and command execution proceeds as before.

Changes of note:
* `cosmovisor/cmd/cosmovisor/cmd/help.go` - New file with the help text and stuff.
* `cosmovisor/args.go` - Updated `GetConfigFromEnv ` and `validate` to check all config pieces instead of returning when the first problem is encountered.
* `cosmovisor/process.go` - Extracted prep into new func `launchCommand` so that it can be used for both `Run` and the new `RunHelp` functions.

Example output:
```console
> ./cosmovisor help
Cosmosvisor - A process manager for Cosmos SDK application binaries.

Cosmovisor is a wrapper for a Cosmos SDK based App (set using the required DAEMON_NAME env variable).
It starts the App by passing all provided arguments and monitors the DAEMON_HOME/data/upgrade-info.json
file to perform an update. The upgrade-info.json file is created by the App x/upgrade module
when the blockchain height reaches an approved upgrade proposal. The file includes data from
the proposal. Cosmovisor interprets that data to perform an update: switch a current binary
and restart the App.

Configuration of Cosmovisor is done through environment variables, which are
documented in: https://github.com/cosmos/cosmos-sdk/tree/master/cosmovisor/README.md

5:04PM INF Configuration is valid:
Configurable Values:
  DAEMON_HOME: /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download
  DAEMON_NAME: foo
  DAEMON_ALLOW_DOWNLOAD_BINARIES: false
  DAEMON_RESTART_AFTER_UPGRADE: true
  DAEMON_POLL_INTERVAL: 300ms
  UNSAFE_SKIP_BACKUP: false
  DAEMON_PREUPGRADE_MAX_RETRIES: 0
Derived Values:
        Root Dir: /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/cosmovisor
     Upgrade Dir: /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/cosmovisor/upgrades
     Genesis Bin: /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/cosmovisor/genesis/bin/foo
  Monitored File: /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/data/upgrade-info.json
 module=cosmovisor
5:04PM ERR  error="current binary is invalid: cannot stat dir /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/cosmovisor/genesis/bin/foo: stat /Users/danielwedul/git/cosmos-sdk/cosmovisor/testdata/download/cosmovisor/genesis/bin/foo: no such file or directory" module=cosmovisor
```

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] ~~added `!` to the type prefix if API or client breaking change~~ (not a breaking change)
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] ~~followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)~~ (not a module change)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] ~~updated the relevant documentation or specification~~ (nothing to update)
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

robert-zaremba commented on cosmovisor/args_test.go:
```diff
@@ -5,18 +5,73 @@ import (
 type argsTestSuite struct {
 	suite.Suite
+
+	envVars []string
```
how about using a map instead of a list? As it is now, it's not very readable - one need to remember the order of env variables.

dwedul-figure replied:
This list is needed no matter what because the tests need to be able to clear out all environment variables involved.

At one point, I had `clearEnv` returning a map and `setEnv` taking in a map, but I didn't like the verbosity and redundancy that was then needed for the test case definitions in `TestGetConfigFromEnv`. I thought about several other solutions, but they all felt like they would be overkill for what is needed.

In the end, I decided that, since it's unit tests and specific to just this file, the `setEnv` helper func could get away with being set up as it is.

robert-zaremba replied:
How about having a struct? This way you will be able to have a short initialization in tests, eg:
```
envVals: CosmovisorEnv{absPath, "testname", "true", "false", "true", "303", "1"}
```
And it will be easy to inspect what each true / false means.

dwedul-figure replied:
I had thought about that, but held off because I didn't think it was that much more helpful that looking at a slice of strings and didn't really want to over-engineer stuff for unit tests. But you're right, it's not the same as looking at a slice of strings, and does make things easier.

I'll make that change.

dwedul-figure replied:
Updated those unit tests to use a `cosmovisorEnv` struct.

I still don't think it was the right move, though. It's now significantly more difficult to maintain the tests when changes are made (specially adding a new environment variable).

Using the struct:
1. Add new field to struct.
1. Add new case to setter.
1. Add new entry to mapper.
1. Update all tests that create a `cosmovisorEnv` to include a value for the new field.
1. Update the couple tests that might have broken due to the actual change.
1. Add new unit tests.

Using the string slices:
1. Add new env var name to `s.envVars`.
1. Update the couple tests that might have broken due to the actual change.
1. Add new unit test cases.

To me, the difficulty in readability of the string slice way is significantly less of a problem than the added difficulty in maintaining the tests. Plus, the more infrastructure you create in order to do unit tests, the more likely you are to introduce conditions where tests should fail, but don't.

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+	if len(args) == 0 {
+		return false
+	}
+	return args[0] == "help" || args[0] == "--help" || (len(args) == 1 && args[0] == "-h")
```
why do we have a special case for `"-h"`?

dwedul-figure replied:
Since `cosmovisor` is a wrapper for another application, I was scared that the other application might define a `-h` that doesn't mean `--help`.

However, I dug through the cobra stuff and found that `--help` and `-h` are added by default. The `help` subcommand is added by default too. Since the Cosmos SDK uses cobra, and `cosmovisor` is designed for apps that implement the Cosmos SDK, I now feel that it's a safe assumption that the `-h` flag will actually be a short form of `--help`.

If any arg is `-h` or `--help` then we should give help. If the first argument is `help`, then we should give help as well.

I limited `help` to only the first argument because, anywhere else (without the dashes), it's either a sub-command, an option value, or an illegal argument. All three are clearly meant for the other application, and in the case of "an option value", it might not actually be requesting help. As such, I didn't want this to drop them into the 2-second timeout, no monitoring version of running the command; doing so might make it impossible for someone to use `cosmovisor` as needed.

I guess, alternatively, I could undo the changes to `process.go` and always have it do the full setup and running of the command, even if we think it might just be for help. And the more I think about that, the more I like it.

dwedul-figure replied:
That last thought grew on me enough that I did it.

If any arg is `help`, `--help`, or `-h`, then help is printed. Then, if the config is valid, the DetailString() is printed and the configured command is run with the provided arguments the same way it would if on of those args weren't present. If the config is not valid, the config errors are printed and execution is stopped.

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+	case nil:
+		// Nothing to do. Move on.
+	case *cosmovisor.MultiError:
+		fmt.Fprintf(os.Stderr, "[cosmovisor] multiple configuration errors found:\n")
```
@spoo-bar  is merging a logger hopefully tomorrow. Let's update this PR after the merge and use a logger instead of `fmt.Fprint`.

dwedul-figure replied:
Since this is part of the help flow, I'm not sure if it's something that needs to be logged, and it's definitely something that should go to either `stderr` or `stdout`.

But I'll take a look once the logger gets merged in.

robert-zaremba replied:
@spoo-bar  merged the logger: https://github.com/cosmos/cosmos-sdk/pull/10217

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+func GetHelpText() string {
+	return fmt.Sprintf(`Cosmosvisor - A process manager for Cosmos SDK application binaries.
+
+Cosmovisor monitors the governance module for incoming chain upgrade proposals.
+If it sees a proposal that gets approved, cosmovisor can automatically download
+the new binary, stop the current binary, switch from the old binary to the new one,
+and finally restart the node with the new binary.
```
Cosmovisor is a wrapper for a Cosmos SDK based App (set using the required
`DAEMON_NAME` env variable). It starts the App by passing all positional
arguments and monitors the `$DAEMON_HOME/data/upgrade-info.json` file to
perform an update. The `upgrade-info.json` is created by the App `x/upgrade`
module when the blockchain height reaches an approved upgrade proposal. The
file includes data from the proposal. Cosmovisor tries to interpret that data
to perform an update: switch a current binary and restarts the App.

robert-zaremba replied:
FYI: cosmovisor doesn't observe `x/gov`. It doesn't stop the App nigher - the App stops by it's own.

dwedul-figure replied:
I had just copied this from the top of the `cosmovisor` readme. I split the first sentence, putting the first part as a short description on the first line, then put the rest as the extra info here.

I've updated it to the text you have in the first comment, though.

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+the new binary, stop the current binary, switch from the old binary to the new one,
+and finally restart the node with the new binary.
+
+Command line arguments are passed on to the configured binary.
+
```
```suggestion
```

dwedul-figure replied:
Done.

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+        Optional, default is 0 (only try the first time, do not retry).
+        The maximum number of times to call pre-upgrade in the application after exit status of 31.
+        After the maximum number of retries, cosmovisor fails the upgrade.
+        Valid values: Integers greater than or equal to 0.
```
I don't think we need to list all the env variables - they are well documented in the README file and I think it's better to keep one source of truth.

dwedul-figure replied:
I pulled this entire section out.

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help.go:
```diff
@@ -0,0 +1,101 @@
+        After the maximum number of retries, cosmovisor fails the upgrade.
+        Valid values: Integers greater than or equal to 0.
+
+For more details, see: https://github.com/cosmos/cosmos-sdk/tree/master/cosmovisor
```
```suggestion
Configuration of Cosmovisor is done through the following environment variables, which 
are documented in: https://github.com/cosmos/cosmos-sdk/tree/master/cosmovisor/README.md
```

dwedul-figure replied:
Updated to:
```
Configuration of Cosmovisor is done through environment variables, which are
documented in: https://github.com/cosmos/cosmos-sdk/tree/master/cosmovisor/README.md
```

robert-zaremba commented on cosmovisor/cmd/cosmovisor/cmd/help_test.go:
```diff
@@ -0,0 +1,153 @@
+type HelpTestSuite struct {
+	suite.Suite
+
+	envVars []string
```
ditto

robert-zaremba commented on cosmovisor/multierror.go:
```diff
@@ -0,0 +1,69 @@
+// String implements the string interface for a MultiError.
+func (e MultiError) String() string {
+	return e.Error()
+}
```
let's put `MultiError` to a new package (eg `errors` - then we should drop `Error` from the name to avoid clashing: `errors.Multi`).
We could also consider moving it the SDK, but until we will restructure SDK, we try to avoid adding it as a dependency to Cosmovisor.

dwedul-figure replied:
Done.

robert-zaremba reviewed (commented):
Good job overall! Few comments:
* I would prefer to update `DetailString` to use same names for the config as the env variables .
* let's move `MultiError` to a separate package

spoo-bar commented on cosmovisor/args.go:
```diff
@@ -265,3 +272,39 @@ func booleanOption(name string, defaultVal bool) (bool, error) {
+		{"RestartAfterUpgrade", fmt.Sprintf("%t", cfg.RestartAfterUpgrade)},
+		{"PollInterval", fmt.Sprintf("%s", cfg.PollInterval)},
+		{"UnsafeSkipBackup", fmt.Sprintf("%t", cfg.UnsafeSkipBackup)},
+	}
```
`{"PreupgradeMaxRetries", fmt.Sprintf("%d", cfg.PreupgradeMaxRetries)},` needed

dwedul-figure replied:
Oops! Thanks! Done.

spoo-bar commented on cosmovisor/args_test.go:
```diff
@@ -168,5 +225,234 @@ func (s *argsTestSuite) TestBooleanOption() {
+		fmt.Sprintf("RestartAfterUpgrade: %t", restartAfterUpgrade),
+		fmt.Sprintf("PollInterval: %s", pollInterval),
+		fmt.Sprintf("UnsafeSkipBackup: %t", unsafeSkipBackup),
+		"Derived Values:",
```
`fmt.Sprintf("PreupgradeMaxRetries: %d", preupgradeMaxRetries)` needed here too

dwedul-figure replied:
Oops! Thanks! Done.

dwedul-figure commented on cosmovisor/args.go:
```diff
@@ -265,3 +272,39 @@ func booleanOption(name string, defaultVal bool) (bool, error) {
+		{"AllowDownloadBinaries", fmt.Sprintf("%t", cfg.AllowDownloadBinaries)},
+		{"RestartAfterUpgrade", fmt.Sprintf("%t", cfg.RestartAfterUpgrade)},
+		{"PollInterval", fmt.Sprintf("%s", cfg.PollInterval)},
+		{"UnsafeSkipBackup", fmt.Sprintf("%t", cfg.UnsafeSkipBackup)},
```
Sounds good. Done.

robert-zaremba reviewed (approved):
ACK. Not blocking, but would be good to take into account:
* logger which is already merged
* use a struct for env argument list (ref: https://github.com/cosmos/cosmos-sdk/pull/10229#discussion_r717525779)

dwedul-figure commented:
I moved the config error handling and output out of the `DoHelp` func into the main `Run` function. I also switched it to use the logger.

I tried out using the logger for the actual help text, but it looked weird and felt it was a little confusing. So I just left it as it was (using `fmt.Println`).

robert-zaremba commented:
Thanks for a great work!
